From ef882bd53a91e48de413c40b69f34f8b3e16266f Mon Sep 17 00:00:00 2001
From: yevgeny.y.rouban <yevgeny.y.rouban@intel.com>
Date: Thu, 7 Mar 2013 18:24:12 +0700
Subject: Dalvik: Fix map from byte code to native code generated

BZ: 73897

mapFromBCtoNCG is expected to map MIR bytecode offsets to offsets of the generated native code
like the following (function lowerByteCode):

   int offsetPC = mir->offset;
   mapFromBCtoNCG[offsetPC] = stream - streamMethodStart

But the function codeGenBasicBlock uses mir->seqNum as index:

   offsetPC = mir->seqNum
   mapFromBCtoNCG[offsetPC] = stream - streamMethodStart

This bug seems caused by overloaded use of name 'offsetPC' for different variables.

Incorrect mapFromBCtoNCG does not affect native code generation but is visible in VTune
source viewer.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I14c27e20ed93e63dc896433bbea03a3c909757ea
Orig-MCG-Change-Id: I40d17c5e164e1384860f033049e41aed593a1a3d
Signed-off-by: yevgeny.y.rouban <yevgeny.y.rouban@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/AnalysisO1.cpp |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/vm/compiler/codegen/x86/AnalysisO1.cpp b/vm/compiler/codegen/x86/AnalysisO1.cpp
index b8b578a..8f3c116 100644
--- a/vm/compiler/codegen/x86/AnalysisO1.cpp
+++ b/vm/compiler/codegen/x86/AnalysisO1.cpp
@@ -1384,7 +1384,7 @@ int codeGenBasicBlock(const Method* method, BasicBlock_O1* bb) {
         } else { //isConst || isDeadStmt
             //if this bytecode is the target of a jump, the mapFromBCtoNCG should be updated
             offsetNCG = stream - streamMethodStart;
-            mapFromBCtoNCG[offsetPC] = offsetNCG;
+            mapFromBCtoNCG[mir->offset] = offsetNCG;
 #ifdef DEBUG_COMPILE_TABLE
             ALOGI("Bytecode %s generates a constant and has no side effect\n", dexGetOpcodeName(mir->dalvikInsn.opcode));
 #endif
-- 
1.7.4.1

